### PR TITLE
Install with pip

### PR DESCRIPTION
**Pull request by musicaljelly**
I can now install using pip by opening a shell at the top level of the project directory, and typing 'pip install .' to tell pip to install whatever's in the current directory.

I've tested wiping my entire jupyter/ipython installation, then running the single command 'pip install .' and this grabs the dependencies as expected before installing our extension.

I have not yet tested installing via the remote repo, but I'll can test that once the pull request gets merged. I expected it to work the same as installing from a local directory.

I changed the readme to specify that (at least to the end user installing our extension), manually copying/symlinking gist.js is no longer necessary.

If you're testing out pip installs, to "uninstall" our extension after doing a pip install, wipe the following directories:

\<PYTHON_INSTALL_DIRECTORY\>/Lib/site-packages/jupyter-notebook-gist
\<PYTHON_INSTALL_DIRECTORY\>/Lib/site-packages/jupyter-notebook-gist-0.0.0-py3.5.egg-info
\<JUPYTER_USER_DIRECTORY\>/nbextensions/gist.js (for me, the full path to this was C:/Users/Lewis/AppData/Roaming/jupyter/nbextensions/gist.js)

There is a 'pip uninstall' command that exists, and we might want to add support for this in the future. Granted, it doesn't look like Jupyter Notebook uninstalls properly when I run this, so I don't know how much people actually expect that command to work. Implementing seems to be considered optional.

There does exists a way to do a pip install that creates symlinks for development use instead of actually copying files. This would let us run 'pip install -e .' once, then never again while we're developing, and the symlink makes sure we're always running the latest version. I haven't played around with this yet. The old method of manually copying gist.js and setting the config settings to point to the right directories should still work for now.

Having a few people test this on different OSs and different versions of python would be good. If you'd like to test, I'd recommend wiping anything in your .jupyter/.ipython config directories that you set up for this extension (other than the github client id/secret, as shown in the new readme). Also wipe the copy of gist.js you copied manually for development. After doing this, run the notebook to make sure the extension does not appear, then do the pip install and confirm that is does.

Please review and let me know of any suggestions/comments/questions!

### CODE REVIEW

**Comment by mreid-moz** [comment]
Great! I will have a look at this tomorrow.

File: README.md [hunk 1]
```diff
@@ -2,38 +2,18 @@
...
-cm.update('notebook', {"load_extensions": {"gist": True}})
-
-# Make the client id *only* available to the client.
 cm.update('notebook', {"oauth_client_id": c.NotebookApp.oauth_client_id})
 ```
 
 Replace the vars above with a working client_id / secret. You can create one
 [here](https://github.com/settings/applications).
 
-Copy or symlink `extensions/gist.js` to your jupyter nbextensions directory.
```
Comment (mreid-moz) [inline README.md]: Please add the instructions for running `pip install .` for now. Once we publish to pip we can update the instructions.

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+from setuptools import setup
```
Comment (mreid-moz) [inline setup.py]: Remove duplicate 'import' of `setup`

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import os
+from setuptools import setup
+from setuptools.command.install import install
+
+EXT_DIR = os.path.join(os.path.dirname(__file__), 'jupyter-notebook-gist')
+
+class InstallCommand(install):
+    def run(self):
+
```
Comment (mreid-moz) [inline setup.py]: nit: remove this blank line

**Comment by musicaljelly** [comment]
Fixed and committed, let me know if you find anything else!

File: setup.py (hunk 3 above)
Comment (mreid-moz) [inline setup.py]: @musicaljelly don't forget this minor style nitpick :)

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        # Activate the Python server extension
+        server_cm = ConfigManager(config_dir=jupyter_config_dir())
+        cfg = server_cm.get('jupyter_notebook_config')
+        server_extensions = cfg.setdefault('NotebookApp', {}).setdefault('server_extensions', [])
+        if "jupyter-notebook-gist.create_gist" not in server_extensions:
+            cfg['NotebookApp']['server_extensions'] += ['jupyter-notebook-gist.create_gist']
+            server_cm.update('jupyter_notebook_config', cfg)
+
+setup(
+    name="jupyter-notebook-gist",
```
Comment (mreid-moz) [inline setup.py]: Please add a `version="0.1.0",` line (or whatever version we want to start from)

Comment (mreid-moz) [inline setup.py]: Oh, and `description="Create a gist from the Jupyter Notebook UI",`

**Comment by mreid-moz** [comment]
Ok, I've tried this out and ran into a couple of problems (possibly due to my wonky local install of jupyter/ipython).

- Hard failure during `pip install .` complaining that the `/Users/mark/.jupyter/nbconfig` directory does not exist. I manually created that dir and was able to proceed.
- Looks like we need to add an `__init__.py` file in the `jupyter-notebook-gist/` directory (an empty one works fine).

**Comment by musicaljelly** [comment]
What error message or exception are you getting when `pip install .` fails? I'm looking through the code in `traitlets/config/manager.py` which handles that config file, and it should create the directory automatically if it doesn't yet exist. This is indeed what happens on my machine if I delete the nbconfig directory and install. Can you repro this consistently when you delete `nbconfig`?

**Comment by mreid-moz** [comment]
Looks like it only fails if the `.jupyter` directory does not exist. I tried it with just a missing `nbconfig` dir and it worked as expected. Probably not worth worrying about - jupyter probably creates that dir when it is installed (and it probably *is* due to my weird configuration).

For what it's worth, here's the error message I see with a missing `.jupyter` dir.

```

  ----------------------------------------
  Failed building wheel for jupyter-notebook-gist
  Running setup.py clean for jupyter-notebook-gist
Failed to build jupyter-notebook-gist
Installing collected packages: jupyter-notebook-gist
  Running setup.py install for jupyter-notebook-gist ... error
    Complete output from command /usr/local/opt/python/bin/python2.7 -u -c "import setuptools, tokenize;__file__='/var/folders/11/11jyx2zd2k75487c1y762hwr0000gn/T/pip-7HrJPA-build/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /var/folders/11/11jyx2zd2k75487c1y762hwr0000gn/T/pip-qA39xE-record/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build/lib
    creating build/lib/jupyter-notebook-gist
    copying jupyter-notebook-gist/__init__.py -> build/lib/jupyter-notebook-gist
    copying jupyter-notebook-gist/create_gist.py -> build/lib/jupyter-notebook-gist
    running install_lib
    creating /usr/local/lib/python2.7/site-packages/jupyter-notebook-gist
    copying build/lib/jupyter-notebook-gist/__init__.py -> /usr/local/lib/python2.7/site-packages/jupyter-notebook-gist
    copying build/lib/jupyter-notebook-gist/create_gist.py -> /usr/local/lib/python2.7/site-packages/jupyter-notebook-gist
    byte-compiling /usr/local/lib/python2.7/site-packages/jupyter-notebook-gist/__init__.py to __init__.pyc
    byte-compiling /usr/local/lib/python2.7/site-packages/jupyter-notebook-gist/create_gist.py to create_gist.pyc
    running install_egg_info
    running egg_info
    writing requirements to jupyter_notebook_gist.egg-info/requires.txt
    writing jupyter_notebook_gist.egg-info/PKG-INFO
    writing top-level names to jupyter_notebook_gist.egg-info/top_level.txt
    writing dependency_links to jupyter_notebook_gist.egg-info/dependency_links.txt
    warning: manifest_maker: standard file '-c' not found

    reading manifest file 'jupyter_notebook_gist.egg-info/SOURCES.txt'
    writing manifest file 'jupyter_notebook_gist.egg-info/SOURCES.txt'
    Copying jupyter_notebook_gist.egg-info to /usr/local/lib/python2.7/site-packages/jupyter_notebook_gist-0.1.0-py2.7.egg-info
    running install_scripts
    writing list of installed files to '/var/folders/11/11jyx2zd2k75487c1y762hwr0000gn/T/pip-qA39xE-record/install-record.txt'
    removing /Users/mark/Library/Jupyter/nbextensions/gist.js
    copying /var/folders/11/11jyx2zd2k75487c1y762hwr0000gn/T/pip-7HrJPA-build/jupyter-notebook-gist/extensions/gist.js -> /Users/mark/Library/Jupyter/nbextensions/gist.js
    error: [Errno 2] No such file or directory: '/Users/mark/.jupyter/nbconfig'

    ----------------------------------------
Command "/usr/local/opt/python/bin/python2.7 -u -c "import setuptools, tokenize;__file__='/var/folders/11/11jyx2zd2k75487c1y762hwr0000gn/T/pip-7HrJPA-build/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /var/folders/11/11jyx2zd2k75487c1y762hwr0000gn/T/pip-qA39xE-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /var/folders/11/11jyx2zd2k75487c1y762hwr0000gn/T/pip-7HrJPA-build
```

**Comment by musicaljelly** [comment]
Gotcha. Yeah that's failing when easy_install.py tries to copy the file, and that directory should definitely exist by then. I tried it with Python 2.7 just to see if that was possibly the difference, but it still worked fine for me. I'll leave that for now, and if anyone else on the team hits it when installing the extension I'll take a closer look.

I've added the __init__.py and the description and version number. Let me know if you find anything further!

**Comment by mreid-moz** [comment]
Looks good now, thanks!

**Pull request merged**